<launch>
	<!-- frobit parameters --> 
	<param name="diff_steer_wheel_distance" value="0.215"/>
	<param name="ticks_per_meter_left" value="562" />
	<param name="ticks_per_meter_right" value="562" />

	<!-- Notice: Use pkg="stage" if running ROS Groovy and pkg="stage_ros" if running ROS Hydro --> 
 	<node pkg="stage_ros" type="stageros" name="simulator" args="$(find sdu_rsd_2014)/stage/frobitpro_robolab_rsd.world"/>
	<remap from="/fmKnowledge/pose" to="/odom" />

 <!-- This assumes roscore is running on this machine -->
	<node pkg="sdu_rsd_waypoint" type="sdu_rsd_waypoint_node" name="waypoint" output="screen">
		<param name="waypoint_sub" value="/fmCommand/waypoint" />
		<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
		<param name="deadman_pub" value="/fmCommand/deadman" />
		<param name="waypointreached_pub" value="/fmKnowledge/waypoint_reached" />
		<param name="odometry_sub" value="/odom" />
		<param name="laserscan_sub" value="/base_scan" />
		<param name="laserscan_frame" value="/base_laser_link" />

		<param name="max_turn_output" value="2" /> <!-- If the output from the turn PID is bigger than this, then it should turn on the spot -->
		<param name="waypoint_reached_threshold" value="0.01" /> <!-- Threshold of when the waypoint is reached. -->
		<param name="max_distance_obstacle" value="0.4" />
		<param name="speed_obstacle" value="0.1" />
		<param name="min_closepoints" value="5" />
		<param name="clearance_distance" value=".5" />

		<param name="kp_obstacle" value="0.1" />
		<param name="kd_obstacle" value="0.0" />

		<param name="kp_angle" value=".1" />
		<param name="ki_angle" value="0.0" />
		<param name="kd_angle" value="0.05" />

		<param name="kp_distance" value="2.4" />
		<param name="ki_distance" value="0.2" />
		<param name="kd_distance" value="1.0" />
	</node>

<group ns="fmExecutors">
		<node pkg="sdu_rsd_2014" type="waypoint_navigation_node.py" name="wptnav" output="screen">
			<param name="automode_sub" value="/fmDecision/automode" />
			<param name="pose_sub" value="/fmKnowledge/pose" />
			<param name="joy_sub" value="/fmLib/joy" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
			<param name="implement_pub" value="/fmCommand/implement" />
			<param name="status_pub" value="/fmInformation/wptnav_status" />
			<param name="pid_pub" value="/fmInformation/wptnav_pid" />
			<param name="status_publish_interval" value="1" />
			<param name="pid_publish_interval" value="1" />
			<param name="print_debug_information" value="false" />

		    	<param name="max_linear_velocity" value="1.2"/>
		    	<param name="max_angular_velocity" value="0.70"/>

		    	<param name="drive_kp" value="2.2"/>
		    	<param name="drive_ki" value="0.5"/>
		    	<param name="drive_kd" value="0.8"/>
		    	<param name="drive_feed_forward" value="0.00"/>
		    	<param name="drive_max_output" value="0.5"/>

		   	<param name="turn_kp" value="1.2"/>
		    	<param name="turn_ki" value="0.8"/>
		    	<param name="turn_kd" value="0.2"/>
		    	<param name="turn_feed_forward" value="0.0"/>
		    	<param name="turn_max_output" value="2.0"/>

			<param name="wpt_default_mode" value="MCTE" />
			<param name="wpt_default_tolerance" value="0.20" />
			<param name="wpt_default_drive_velocity" value="0.4" />
			<param name="wpt_default_turn_velocity" value="1.4" />
			<param name="wpt_default_wait_after_arrival" value="2.0" />
			<param name="wpt_default_implement_command" value="0.0" />

			<param name="target_ahead" value="0.3" />
			<param name="turn_start_at_heading_err" value="60.0" />
			<param name="turn_stop_at_heading_err" value="2.0" />
			<param name="ramp_drive_velocity_at_distance" value="0.25" />
			<param name="ramp_min_drive_velocity" value="0.10" />
			<param name="ramp_turn_velocity_at_angle" value="30.0" />
			<param name="ramp_min_turn_velocity" value="0.3" />
			<param name="stop_navigating_at_distance" value="0.03" />
		</node>		
	</group>

<group ns="fmLib">
		<node pkg="twist_stamped_to_twist" name="converter" type="cmd_vel_converter.py">	
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="cmd_vel_pub" value="/cmd_vel" />
			<param name="deadman_sub" value="/fmCommand/deadman" />
		</node>
</group>
</launch>
